01/02/2021

Antonio Pascual Hernández

# TAREA FINAL: Datos de vivienda

<p> El negocio inmobiliario es uno de los motores de la economía de España, el carácter turístico de nuesta península hace que sean muchos los inversores extranjeros que se decidan por buscar una inversión de bajo riesgos en nuestras ciudades costeras o grandes urbes. 
Muchas de estas inversiones se hacen sobre grandes bolsas de inmuebles que deben ser analizados previamente para comprobar la rentabilidad del porfolio </p>

<!-- <p> En este caso vamos a trabajar con una tabla que contienen información de distintos inmuebles repartidos por una zona específica, sus carácterísticas y su precio </p>  -->

Todas las cuestiones se deben realizar sobre el conjunto de casos que representen viviendas ubicadas en zonas residenciales **(alta, media y baja densidad)**

**MUY IMPORTANTE:** En las otras prácticas he detectado colaboraciones involucrando varias personas y he sido flexible aunque a algunos os lo he mencionado en las correcciones, porque al final el trabajo de analista de datos es un trabajo colaborativo. Sin embargo, este trabajo es individual, así que cuidado con las colaboraciones.

Las variables de las que se compone el dataset son:

|NOMBRE VARIABLE|DESCRIPTOR|VALORES|
| --- | --- | --- |
|Order|Variable de identificación|1 a 2930|
|MS Zoning|Zona de ubicación de la vivienda|"A rural, C comercial, FV residencial flotante, I industrial, RH residencial alta densidad, RL residencial baja densidad, RM residencial media densidad"|
|Lot Frontage|Longitud de la fachada en pies||
|Lot Area|Superficie de la vivienda en pies cuadrados||
|Land Contour|Contorno del terreno circundante|"Lvl llano, Bnk Tipo bancal, HLS Ladera, Low Depresión"|
|Land Slope|Tipo de pendiente de la vivienda|" Gtl pendiente suave, Mod pendiente moderada, Sev fuerte pendiente"|
|Overall Qual|Grado de calidad de materiales y acabado de la vivienda|De 1 (Muy pobre) a 10 (Excelente)|
|Year Built|Año de construccion de la vivienda||
|Year Remod/Add|Año de última reforma de la vivienda||
|Mas Vnr Type|Tipo de revestimiento exterior|" BrkCmn Ladrillo normal, BrkFace Ladrillo visto, CBlock Bloque de cemento, None Ninguna, Stone Piedra "|
|Exter Qual|Calidad de revestimiento exterior|"Ex Excelente,Gd Bueno,TA Media,Fa Justo"|
|Bsmt Cond|Estado general del sótano|"Ex Excelente, Gd Bueno, TA Media, Fa Justo, Po Pobre,Ss sin sótano"|
|Total Bsmt SF|Superficie del sótano en pies cuadrados|
|Heating QC|Calidad de la calefacción|"Ex Excelente,Gd Bueno,TA Media,Fa Justo,Po Pobre"|
|Central Air|Aire acondicionado centralizado|"N No Y Sí"|
|Full Bath|Número de baños completo en planta||
|Half Bath|Número de aseos en planta||
|Bedroom AbvGr|Número de dormitorios en planta||
|Kitchen AbvGr|Número de cocinas en planta||
|Kitchen Qual|Calidad de cocinas|"Ex Excelente,Gd Bueno,TA Media,Fa Justo,Po Pobre"|
|TotRms AbvGrd|Número total de habitaciones excluidos los cuartos de baño||
|Garage Cars|Número de plazas de garaje||
|Garage Area|Superficie del garaje|||
|Garage Cond|Estado del garaje|"Ex Excelente,Gd Bueno,TA Media,Fa Justo,Po Pobre,Sg sin garaje"|
|Pool Area|Superficie de la piscina en pies cuadrados|
|Pool QC|Calidad de la piscina|"Ex Excelente,Gd Bueno,TA Media,Fa Justo,Sp no hay piscina"|
|Mo Sold|mes de venta||
|Yr Sold|año de venta||
|SalePrice|precio de venta en dólares||

Recomiendo al leer los datos, eliminar espacios de los nombres de las columnas, realiza un pequeño análisis inicial de los mismos. No olvides fijarte en los tipos de variables, que variables pueden tener tipos confundidos y corrige los. Sobre todo, trabaja con las fechas.

**NOTA:** Las tareas complementarias sirven para subir nota. El resto de preguntas valen igual y suman 10 puntos.

## Inicializar y cargar el contexto spark

In [12]:
import pandas as pd #se carga pandas
import numpy as np #se carga numpy

In [1]:
!apt-get install openjdk-8-jdk-headless -qq > /dev/null
!wget -q https://www-us.apache.org/dist/spark/spark-2.4.7/spark-2.4.7-bin-hadoop2.7.tgz
!tar xf spark-2.4.7-bin-hadoop2.7.tgz #d
!pip install -q findspark 

import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
os.environ["SPARK_HOME"] = "/content/spark-2.4.7-bin-hadoop2.7"

import findspark
findspark.init()
from pyspark import SparkContext
sc = SparkContext()
from pyspark.sql import SparkSession
spark = SparkSession.builder.master("local[*]").getOrCreate()

In [2]:
vivienda = sc.textFile("BDpracticafinalCSV.csv") #se importan los datos
vivienda = vivienda.map(lambda x: x.split(";")) #separamos por el punto y coma
header = vivienda.first()
vivienda = vivienda.filter(lambda x: x != header) #eliminamos la primera fila, que hace referencia al encabezado
vivienda.take(5)

[['1',
  'RL',
  '141',
  '31770',
  'Lvl',
  'Gtl',
  '6',
  '1960',
  '1960',
  'Stone',
  'TA',
  'Gd',
  '1080',
  'Fa',
  'Y',
  '1',
  '0',
  '3',
  '1',
  'TA',
  '7',
  '2',
  '528',
  'TA',
  '0',
  'Sp',
  '5',
  '2010',
  '215000'],
 ['2',
  'RH',
  '80',
  '11622',
  'Lvl',
  'Gtl',
  '5',
  '1961',
  '1961',
  'None',
  'TA',
  'TA',
  '882',
  'TA',
  'Y',
  '1',
  '0',
  '2',
  '1',
  'TA',
  '5',
  '1',
  '730',
  'TA',
  '0',
  'Sp',
  '6',
  '2010',
  '105000'],
 ['3',
  'RL',
  '81',
  '14267',
  'Lvl',
  'Gtl',
  '6',
  '1958',
  '1958',
  'BrkFace',
  'TA',
  'TA',
  '1329',
  'TA',
  'Y',
  '1',
  '1',
  '3',
  '1',
  'Gd',
  '6',
  '1',
  '312',
  'TA',
  '0',
  'Sp',
  '6',
  '2010',
  '172000'],
 ['4',
  'RL',
  '93',
  '11160',
  'Lvl',
  'Gtl',
  '7',
  '1968',
  '1968',
  'None',
  'Gd',
  'TA',
  '2110',
  'Ex',
  'Y',
  '2',
  '1',
  '3',
  '1',
  'Ex',
  '8',
  '2',
  '522',
  'TA',
  '0',
  'Sp',
  '4',
  '2010',
  '244000'],
 ['5',
  'RL',
  '74',
  '13

In [13]:
pd_vivienda = pd.read_csv("BDpracticafinalCSV.csv", delimiter = ';') #se carga el dataset con pandas
pd_vivienda.head(5)

Order MS Zoning  Lot Frontage  Lot Area  ... Pool QC Mo Sold  Yr Sold  SalePrice
0      1        RL         141.0     31770  ...      Sp       5     2010     215000
1      2        RH          80.0     11622  ...      Sp       6     2010     105000
2      3        RL          81.0     14267  ...      Sp       6     2010     172000
3      4        RL          93.0     11160  ...      Sp       4     2010     244000
4      5        RL          74.0     13830  ...      Sp       3     2010     189900

[5 rows x 29 columns]

A continución se seleccionarán las observaciones correspondientes a viviendas ubicadas en zonas residenciales (alta, media y baja densidad).

In [ ]:
pd_vivienda["MS Zoning"].unique()

En cuanto a las observaciones de las zonas residenciales de baja densidad, se encuentran 3 opciones distintas (categorías): RL, Rl y rL. Por lo tanto, se deberá tener en cuenta a la hora de seleccionar las variables.

In [3]:
bajo = vivienda.filter(lambda x: x[1] == 'RL') #filtro que selecciona las zonas residenciales de baja densidad
alto = vivienda.filter(lambda x: x[1] == 'RH') #filtro que selecciona las zonas residenciales de alta densidad
medio = vivienda.filter(lambda x: x[1] == 'RM') #filtro que selecciona las zonas residenciales de media densidad
bajo1 = vivienda.filter(lambda x: x[1] == 'Rl')
bajo2 = vivienda.filter(lambda x: x[1] == 'rL')

Se unen los diferentes objetos de las zonas, y se observa como ha quedado brevemente.

In [4]:
residencia = bajo.union(alto) #se unen las observaciones obtenidas de los filtros anteriores
residencia2 = residencia.union(medio)
residencia3 = residencia2.union(bajo1)
vivienda = residencia3.union(bajo2)

## Cuántas viviendas distintas encontramos en el dataset? ¿Se repite alguna? Tiene sentido que haya duplicadas? ¿Qué podemos hacer con las duplicadas?

In [ ]:
vivienda.count() #Se utiliza la función count() para obtener el número de viviendas.

2768

In [ ]:
len(pd_vivienda)-len(pd_vivienda.drop_duplicates()) #se utiliza pandas para obtener los duplicados

6

Se obtienen 6 duplicados del dataset completo. 
Si podría tener sentido que hubiese duplicados. Se podría pensar que existan viviendas de una misma urbanización o edificio, lo que implicaría que sus caracterísitcas podrían ser iguales.

## ¿Podrías decirme el total de inmuebles y el precio medio (Sale Price) de cada zona (MS Zoning)?

Baja densidad:

In [ ]:
baja1 = bajo.union(bajo1)
baja2 = baja1.union(bajo2)

In [ ]:
baja_densidad = baja2.map(lambda x: (x[28]))
baja_densidad = baja_densidad.filter(lambda x_int: x_int.isdigit())
baja_densidad = baja_densidad.map(lambda x: float(x))
baja_densidad.mean(), baja_densidad.count()

(191412.6068451076, 2279)

Viviendas de baja densidad: 2279

Precio media de baja densidad: 191412.61$

Media densidad:

In [ ]:
media_densidad = medio.map(lambda x: (x[28]))
media_densidad = media_densidad.filter(lambda x_int: x_int.isdigit())
media_densidad = media_densidad.map(lambda x: float(x))
media_densidad.mean(), media_densidad.count()

(126781.39393939397, 462)

Viviendas de media densidad: 462

Precio media de media densidad: 126781.39$

Zona RH:

In [ ]:
alta_densidad = alto.map(lambda x: (x[28]))
alta_densidad = alta_densidad.filter(lambda x_int: x_int.isdigit())
alta_densidad = alta_densidad.map(lambda x: float(x))

alta_densidad.mean(), alta_densidad.count()

(136419.77777777778, 27)

Viviendas de media densidad: 27

Precio media de media densidad: 136419.78$

## Media de Total Bsmt SF por cada década de construcción calculada a partir de Year Built. 
## ¿Cuál es la decada de construcción con viviendas mejor acondicionadas para el frío (Heating QC)?

In [ ]:
def get_decade(x):
    return x[0] + x[1] + x[2] + "0s" #se define la función que obtendrá las decadas. Como podrías coincidir decadas si solo tenemos en cuenta los dos últimos digitos, se seleccionan todos los digitos del numero (1980 = 80s, 1880 = 80s )

In [ ]:
excelente = vivienda.filter(lambda x: (x[13]== "Ex")) #se seleccionan las observaciones cuyas condiciones de acondicionamiento sean excelentes
excelente = excelente.map(lambda x: ( x[13], get_decade(x[7])))
excelente = excelente.map( lambda x: (x[1],x[0]))
excelente.countByKey()

defaultdict(int,
            {'1880s': 2,
             '1890s': 4,
             '1900s': 12,
             '1910s': 38,
             '1920s': 51,
             '1930s': 40,
             '1940s': 56,
             '1950s': 98,
             '1960s': 88,
             '1970s': 68,
             '1980s': 28,
             '1990s': 235,
             '2000s': 641,
             '2010s': 2})

La década con viviendas mejor acondicionadas es la década de los 2000s

## ¿Cuáles son las 10 viviendas que se vendieron por un precio más elevado por metro cuadrado en el año 2009?

In [ ]:
venta2009 = vivienda.filter(lambda x: x[27] == '2009') #se seleccionan las viviendas que se vendieron en 2009

mascaras = venta2009.map(lambda x: (x[27], [float(x[28]), float(x[3]) * 0.09290304])) #como un pie cuadrado = 0.09290304 metros cuadrados, se multiplicarán los datos de la columna "Lot Area" por ese valor

mascaras.map(lambda x: [x[0],x[1][0]/x[1][1]]).top(10) #listado de las 10 viviendas más caras por metro cuadrado

[['2009', 1026.7114551323118],
 ['2009', 991.0320746567418],
 ['2009', 821.056658770186],
 ['2009', 807.9610176314369],
 ['2009', 765.6471992838166],
 ['2009', 756.0365649831829],
 ['2009', 717.5940277806482],
 ['2009', 714.2814183044536],
 ['2009', 711.1869382468924],
 ['2009', 701.0704810883306]]

## Media anual por zonas del precio de venta y metros cuadrados.

Media precio de venta en zona RM:

In [ ]:
pd_vivienda['Yr Sold'].unique()

array([2010, 2009, 2008, 2007, 2006])

Para las zonas de baja densidad (RL):

Se mostrarán 5 grupos de datos (desde 2010 a 2006), donde el primer dato indicará la media del precio de venta, y el segundo indica la media de los metros cuadrados.

In [ ]:
baja = bajo.union(bajo1)
bajadensidad = baja.union(bajo2)

In [ ]:
RLventa2010 = bajadensidad.filter(lambda x: x[27] == '2010')
RLventa2009 = bajadensidad.filter(lambda x: x[27] == '2009')
RLventa2008 = bajadensidad.filter(lambda x: x[27] == '2008')
RLventa2007 = bajadensidad.filter(lambda x: x[27] == '2007')
RLventa2006 = bajadensidad.filter(lambda x: x[27] == '2006')

In [ ]:
precioRL2010 = RLventa2010.map(lambda x: int(x[28]))
metrosRL2010 = RLventa2010.map(lambda x: float(x[3]) * 0.09290304)
precioRL2010.mean(), metrosRL2010.mean()

(184978.41897233203, 1011.35975210751)

In [ ]:
precioRL2009 = RLventa2009.map(lambda x: int(x[28]))
metrosRL2009 = RLventa2009.map(lambda x: float(x[3]) * 0.09290304)
precioRL2009.mean(), metrosRL2009.mean()

(190455.5508637236, 1002.1845004873709)

In [ ]:
precioRL2008 = RLventa2008.map(lambda x: int(x[28]))
metrosRL2008 = RLventa2008.map(lambda x: float(x[3]) * 0.09290304)
precioRL2008.mean(), metrosRL2008.mean()

(189885.5851528385, 1047.653788615546)

In [ ]:
precioRL2007 = RLventa2007.map(lambda x: int(x[28]))
metrosRL2007 = RLventa2007.map(lambda x: float(x[3]) * 0.09290304)
precioRL2007.mean(), metrosRL2007.mean()

(193399.98014440428, 1059.6784023578332)

In [ ]:
precioRL2006 = RLventa2006.map(lambda x: int(x[28]))
metrosRL2006 = RLventa2006.map(lambda x: float(x[3]) * 0.09290304)
precioRL2006.mean(), metrosRL2006.mean()

(194911.2819472616, 1043.6889575698176)

Para las zonas de media densidad (RM):

Se mostrarán 5 grupos de datos (desde 2010 a 2016), donde el primer dato indicará la media del precio de venta, y el segundo indica la media de los metros cuadrados.

In [ ]:
RMventa2010 = medio.filter(lambda x: x[27] == '2010')
RMventa2009 = medio.filter(lambda x: x[27] == '2009')
RMventa2008 = medio.filter(lambda x: x[27] == '2008')
RMventa2007 = medio.filter(lambda x: x[27] == '2007')
RMventa2006 = medio.filter(lambda x: x[27] == '2006')

In [ ]:
precioRM2010 = RMventa2010.map(lambda x: int(x[28]))
metrosRM2010 = RMventa2010.map(lambda x: float(x[3]) * 0.09290304)
precioRM2010.mean(), metrosRM2010.mean()

(119681.70689655174, 586.1044562648276)

In [ ]:
precioRM2009 = RMventa2009.map(lambda x: int(x[28]))
metrosRM2009 = RMventa2009.map(lambda x: float(x[3]) * 0.09290304)
precioRM2009.mean(), metrosRM2009.mean()

(118759.86904761905, 552.5563142400001)

In [ ]:
precioRM2008 = RMventa2008.map(lambda x: int(x[28]))
metrosRM2008 = RMventa2008.map(lambda x: float(x[3]) * 0.09290304)
precioRM2008.mean(), metrosRM2008.mean()

(133336.1219512195, 623.6694371590244)

In [ ]:
precioRM2007 = RMventa2007.map(lambda x: int(x[28]))
metrosRM2007 = RMventa2007.map(lambda x: float(x[3]) * 0.09290304)
precioRM2007.mean(), metrosRM2007.mean()

(128947.62105263153, 625.4486913751578)

In [ ]:
precioRM2006 = RMventa2006.map(lambda x: int(x[28]))
metrosRM2006 = RMventa2006.map(lambda x: float(x[3]) * 0.09290304)
precioRM2006.mean(), metrosRM2006.mean()

(127502.63725490194, 575.0142579388239)

Para las zonas de alta densidad (RH):

Se mostrarán 5 grupos de datos (desde 2010 a 2016), donde el primer dato indicará la media del precio de venta, y el segundo indica la media de los metros cuadrados.

In [ ]:
RHventa2010 = alto.filter(lambda x: x[27] == '2010')
RHventa2009 = alto.filter(lambda x: x[27] == '2009')
RHventa2008 = alto.filter(lambda x: x[27] == '2008')
RHventa2007 = alto.filter(lambda x: x[27] == '2007')
RHventa2006 = alto.filter(lambda x: x[27] == '2006')

In [ ]:
precioRH2010 = RHventa2010.map(lambda x: int(x[28]))
metrosRH2010 = RHventa2010.map(lambda x: float(x[3]) * 0.09290304)
precioRH2010.mean(), metrosRH2010.mean()

(126280.0, 753.0348810240001)

In [ ]:
precioRH2009 = RHventa2009.map(lambda x: int(x[28]))
metrosRH2009 = RHventa2009.map(lambda x: float(x[3]) * 0.09290304)
precioRH2009.mean(), metrosRH2009.mean()

(116206.8, 647.143996032)

In [ ]:
precioRH2008 = RHventa2008.map(lambda x: int(x[28]))
metrosRH2008 = RHventa2008.map(lambda x: float(x[3]) * 0.09290304)
precioRH2008.mean(), metrosRH2008.mean()

(150333.33333333334, 766.1404032)

In [ ]:
precioRH2007 = RHventa2007.map(lambda x: int(x[28]))
metrosRH2007 = RHventa2007.map(lambda x: float(x[3]) * 0.09290304)
precioRH2007.mean(), metrosRH2007.mean()

(147250.0, 651.31998768)

In [ ]:
precioRH2006 = RHventa2006.map(lambda x: int(x[28]))
metrosRH2006 = RHventa2006.map(lambda x: float(x[3]) * 0.09290304)
precioRH2006.mean(), metrosRH2006.mean()

(140316.66666666666, 659.56513248)

## ¿Podrías decirme el total de recaudación de las casas de revistimiento (Mas Vnr Type) de piedra con respecto a las de ladrillo? ¿Hay diferencia significativa?

In [ ]:
pd_vivienda['Mas Vnr Type'].unique()

array(['Stone', 'None', 'BrkFace', nan, 'BrkCmn', 'CBlock'], dtype=object)

In [ ]:
stone = vivienda.filter(lambda x: x[9] == 'Stone') #se seleccionan las viviendas con revestimiento de piedra
brick = vivienda.filter(lambda x: x[9] == 'BrkFace' or 'BrkCmn') #se seleccionan las viviendas con revestimiento de ladrillo

In [ ]:
stone = stone.map(lambda x: (x[28]))
stone = stone.filter(lambda x_int: x_int.isdigit())
stone = stone.map(lambda x: int(x))
stone.sum(), stone.count()

(60495477, 230)

In [ ]:
brick = brick.map(lambda x: (x[28]))
brick = brick.filter(lambda x_int: x_int.isdigit())
brick = brick.map(lambda x: int(x))
brick.sum(), brick.count()

(498485669, 2768)

Hay una diferencia significativa entre el total de recaudación de las viviendas de de revestimiento de piedra y revestimiento de ladrillo. Esto se podría deber a que existen muchas más viviendas con revistimiento de ladrillo (2768) que viviendas con revestimiento de piedra (230).

## ¿Cuánto son más caras las viviendas con 2 cocinas, con 2 o más plazas de garaje que las que tienen 1 cocina y 1 plaza de garaje? Comparar medias y cuartiles de ambos casos

In [ ]:
doscocinas = vivienda.filter(lambda x: x[18] == '2' and x [21] == '2' or '3' or '4' or '5') #se seleccionan las viviendas con 2 cocinas y 2 o más plazas de garaje
doscocinas = doscocinas.map(lambda x: (x[28]))
doscocinas = doscocinas.filter(lambda x_int: x_int.isdigit())
doscocinas = doscocinas.map(lambda x: int(x))

In [ ]:
unacocina = vivienda.filter(lambda x: x[18] == '1' and x [21] == '1') #se seleccionan las viviendas con 1 cocina y 1 plaza de garaje
unacocina = unacocina.map(lambda x: (x[28]))
unacocina = unacocina.filter(lambda x_int: x_int.isdigit())
unacocina = unacocina.map(lambda x: int(x))

In [ ]:
doscocinas.mean() - unacocina.mean()

51966.75987396165

Si se comparan las medias del precio de las viviendas con 2 cocinas y con 2 o más plazas de garaje con la media de las viviendas con una cocina y 1 plaza de garaje, se observa que el primer tipo de viviendas son,de media, 51966.76$ más caras que las viviendas del segundo tipo.

## (COMPLEMENTARIA) Estudiar la relación entre el precio y el número de Garajes. (Recomiendo segmentar precio por cuartiles y estudiar el número de casos coincidentes)

In [14]:
pd_vivienda['Garage Cars'].unique()

array([ 2.,  1.,  3.,  0.,  4.,  5., nan])

In [17]:
nogarage = vivienda.filter(lambda x: (x[21] == "0")).map(lambda x: (int(x[28])))
unaplaza = vivienda.filter(lambda x: (x[21]== "1")).map(lambda x: (int(x[28])))
dosplazas = vivienda.filter(lambda x: (x[21]== "2")).map(lambda x: (int(x[28])))
tresplazas = vivienda.filter(lambda x: (x[21]== "3")).map(lambda x: (int(x[28])))
cuatroplazas = vivienda.filter(lambda x: (x[21]== "4")).map(lambda x: (int(x[28])))
cincoplazas = vivienda.filter(lambda x: (x[21]== "5")).map(lambda x: (int(x[28])))

In [20]:
nogarage.mean(), nogarage.count()

(107023.63945578231, 147)

In [21]:
unaplaza.mean(), unaplaza.count()

(128011.65575916233, 764)

In [19]:
dosplazas.mean(), dosplazas.count()

(181890.0931802836, 1481)

In [22]:
tresplazas.mean(), tresplazas.count()

(311831.48882681574, 358)

In [23]:
cuatroplazas.mean(), cuatroplazas.count()

(228748.6875, 16)

In [24]:
cincoplazas.mean(), cincoplazas.count()

(126500.0, 1)

No se puede establecer una relación clara entre el precio de las viviendas y el número de plazas de garajes. Lo normal sería pensar que a más plazas de garaje, mayor sería el precio de la vivienda, pero si se comparan los pecio medios de las viviendas por plazas de garaje, se observa que el precio va aumentando hasta las tres plazas, pero el precio medio de las viviendas con 4 plazas es menor.


## (COMPLEMENTARIA) Las 10 viviendas con mejores servicios y mejor precio.

### Pista:
#### Calcula las variables: 
- Número de servicios excelentes
- Número de servicios buenos
...

#### Tendréis que tener en cuenta también variables como número de baños Full Bath, cocinas Kitchen AbvGr o dormitorios Bedroom AbvGr. Generando por ejemplo (número de estas variables por encima de media)

#### Finalmente precio de venta

In [11]:
mejor = vivienda.filter(lambda x: (x[6] in ('8', '9', '10'))) #grado de calidad de los materiales
mejor1 = mejor.filter(lambda x: (x[10] in ("Ex", 'Gd'))) #calidad del revestimiento
mejor2 = mejor1.filter(lambda x: (x[13] in ("Ex", 'Gd'))) #calidad de calefacción
mejor3 = mejor2.filter(lambda x: (x[19] in ("Ex", 'Gd'))) #calidad cocina

mejores = mejor3.map(lambda x: (float(x[28]),x[7]))

mejores.takeOrdered(10, key = lambda x: x[0])

[(147000.0, '2006'),
 (160000.0, '2008'),
 (162500.0, '1989'),
 (168000.0, '1882'),
 (170000.0, '1986'),
 (178000.0, '2004'),
 (179900.0, '2007'),
 (180000.0, '1993'),
 (180000.0, '1992'),
 (180500.0, '2001')]